core/src/enc/bases.rs

```diff
@@ -0,0 +1,201 @@
+//! Contains simplified access to `bech32` and `base58check` encoder/decoder for Bitcoin
+//! addresses. Also defines common encoder errors.
+
+use bech32::{
+    decode as b32_decode, encode as b32_encode, u5, Error as BechError, FromBase32, ToBase32,
+};
+
+use base58check::{FromBase58Check, FromBase58CheckError, ToBase58Check};
+
+use thiserror::Error;
+
+/// Errors that can be returned by the Bitcoin `AddressEncoder`.
+#[derive(Debug, Error)]
+pub enum EncodingError {
+    /// Returned when ScriptPubkey type is unknown. May be non-standard or newer than lib version.
+    #[error("Non-standard ScriptPubkey type")]
+    UnknownScriptType,
+
+    /// Bech32 HRP does not match the current network.
+    #[error("Bech32 HRP does not match. \nGot {:?} expected {:?} Hint: Is this address for another network?", got, expected)]
+    WrongHRP {
+        /// The actual HRP.
+        got: String,
+        /// The expected HRP.
+        expected: String,
+    },
+
+    /// Base58Check version does not match the current network
+    #[error("Base58Check version does not match. \nGot {:?} expected {:?} Hint: Is this address for another network?", got, expected)]
+    WrongVersion {
+        /// The actual version byte.
+        got: u8,
+        /// The expected version byte.
+        expected: u8,
+    },
+
+    /// Bubbled up error from base58check library
+    #[error("FromBase58CheckError: {:?}", .0)]
+    B58Error(FromBase58CheckError),
+
+    /// Bubbled up error from bech32 library
+    #[error("BechError: {:?}", .0)]
+    BechError(#[from] BechError),
+
+    /// Op Return ScriptPubkey was passed to encoder
+    #[error("Can't encode op return scripts as addresses")]
+    NullDataScript,
+}
+
+/// Impl explicitly because FromBase58CheckError doesn't implement the std error format
+impl From<FromBase58CheckError> for EncodingError {
+    fn from(e: FromBase58CheckError) -> Self {
+        EncodingError::B58Error(e)
+    }
+}
+
+/// A simple result type alias
+pub type EncodingResult<T> = Result<T, EncodingError>;
+
+/// Encode a byte vector to bech32. This function expects `v` to be a witness program, and will
+/// return an `UnknownScriptType` if it does not meet the witness program format.
+pub fn encode_bech32(hrp: &str, v: u8, h: &[u8]) -> EncodingResult<String> {
```

Comment:
I changed the interface to `encode_bech32` to take the hrp, version and hash (or data with segwit v1). The network specific wrapper is responsible for being able to get the version.

Reply:
:+1: